core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -1165,16 +1189,23 @@ class KafkaApis(val requestChannel: RequestChannel,
     }
 
     val unauthorizedForCreateTopicMetadata = unauthorizedForCreateTopics.map(topic =>
-      metadataResponseTopic(Errors.TOPIC_AUTHORIZATION_FAILED, topic, isInternal(topic), util.Collections.emptyList()))
+      // Set topicId to zero since we will never create topic which topicId
+      metadataResponseTopic(Errors.TOPIC_AUTHORIZATION_FAILED, topic, Uuid.ZERO_UUID, isInternal(topic), util.Collections.emptyList()))
 
     // do not disclose the existence of topics unauthorized for Describe, so we've not even checked if they exist or not
     val unauthorizedForDescribeTopicMetadata =
       // In case of all topics, don't include topics unauthorized for Describe
       if ((requestVersion == 0 && (metadataRequest.topics == null || metadataRequest.topics.isEmpty)) || metadataRequest.isAllTopics)
         Set.empty[MetadataResponseTopic]
-      else
+      else if (useTopicId) {
+        // We should not return information about existence of a topic on unauthorized error, so we return an UNKNOWN_TOPIC_ID
+        unauthorizedForDescribeTopics.map(topic =>
+          metadataResponseTopic(Errors.UNKNOWN_TOPIC_ID, null, metadataCache.getTopicId(topic), false, util.Collections.emptyList()))
```

Comment:
In the time since we last looked at this, we decided that it is ok to acknowledge existence of the topic ID and return TOPIC_AUTHORIZATION_FAILED. This is so the user knows it is not a retriable error. Please see https://issues.apache.org/jira/browse/KAFKA-12394 for more details.